src/storage/recovery/recovery_manager.cpp

```diff
@@ -0,0 +1,934 @@
+#include <algorithm>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+#include "storage/recovery/recovery_manager.h"
+
+#include "catalog/postgres/pg_attribute.h"
+#include "catalog/postgres/pg_class.h"
+#include "catalog/postgres/pg_constraint.h"
+#include "catalog/postgres/pg_database.h"
+#include "catalog/postgres/pg_index.h"
+#include "catalog/postgres/pg_namespace.h"
+#include "catalog/postgres/pg_type.h"
+#include "storage/index/index_builder.h"
+#include "storage/write_ahead_log/log_io.h"
+
+namespace terrier::storage {
+
+uint32_t RecoveryManager::RecoverFromLogs() {
+  // Replay logs until the log provider no longer gives us logs
+  uint32_t txns_replayed = 0;
+  while (true) {
+    auto pair = log_provider_->GetNextRecord();
+    auto *log_record = pair.first;
+
+    // If we have exhausted all the logs, break from the loop
+    if (log_record == nullptr) break;
+
+    switch (log_record->RecordType()) {
+      case (LogRecordType::ABORT): {
+        TERRIER_ASSERT(pair.second.empty(), "Abort records should not have any varlen pointers");
+        ProcessAbortedTransaction(log_record->TxnBegin());
+        delete[] reinterpret_cast<byte *>(log_record);
+        break;
+      }
+
+      case (LogRecordType::COMMIT): {
+        TERRIER_ASSERT(pair.second.empty(), "Commit records should not have any varlen pointers");
+        auto *commit_record = log_record->GetUnderlyingRecordBodyAs<CommitRecord>();
+
+        // We defer all transactions
+        DeferTransaction(log_record->TxnBegin());
+
+        // Process any deferred transactions that are safe to execute
+        txns_replayed += ProcessDeferredTransactions(commit_record->OldestActiveTxn());
+
+        // Clean up the log record
+        delete[] reinterpret_cast<byte *>(log_record);
+        break;
+      }
+
+      default:
+        TERRIER_ASSERT(
+            log_record->RecordType() == LogRecordType::REDO || log_record->RecordType() == LogRecordType::DELETE,
+            "We should only buffer changes for redo or delete records");
+        buffered_changes_map_[log_record->TxnBegin()].push_back(pair);
+    }
+  }
+  // Process all deferred txns
+  ProcessDeferredTransactions(transaction::NO_ACTIVE_TXN);
+  TERRIER_ASSERT(deferred_txns_.empty(), "We should have no unprocessed deferred transactions at the end of recovery");
+
+  // If we have unprocessed buffered changes, then these transactions were in-process at the time of system shutdown.
+  // They are unrecoverable, so we need to clean up the memory of their records.
+  if (!buffered_changes_map_.empty()) {
+    for (auto &txn : buffered_changes_map_) {
```

Comment:
I would definitely defer this one to the GC.  At this point, we want to exit as fast as possible because we are "recovered" and can now go live and start accepting write transactions.